keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,232 @@
+            - [Trace Utility Package](#trace-utility-package)
+            - [Tracing Pod Lifecycle](#tracing-pod-lifecycle)
+        - [Out-of-tree changes](#out-of-tree-changes)
+            - [Tracing best-practices documentation](#tracing-best-practices-documentation)
+            - [Mutating admission webhook](#mutating-admission-webhook)
+    - [Graduation Requirements](#graduation-requirements)
+    - [Implementation History](#implementation-history)
+
+## Summary
+
+This Kubernetes Enhancement Proposal (KEP) introduces a model for adding distributed tracing to Kubernetes object lifecycles. The inclusion of this trace instrumentation will mark a significant step in making Kubernetes processes more observable, understandable, and debuggable.
+
+
+## Motivation
+
+Debugging latency issues in Kubernetes is an involved process. There are existing tools which can be used to isolate these issues in Kubernetes, but these methods fall short for various reasons. For instance:
+
+* **Logs**: are fragmented, and finding out which process was the bottleneck involves digging through troves of unstructured text. In addition, logs do not offer higher-level insight into overall system behavior without an extensive background on the process of interest. 
+* **Events**: in Kubernetes are only kept for an hour by default, and don't integrate with visualization of analysis tools. To gain trace-like insights would require a large investment in custom tooling.
+* **Latency metrics**: can only supply limited metadata because of cardinality constraints.  They are useful for showing _that_ a process was slow, but don't provide insight into _why_ it was slow.
+* **Latency Logging**: is a "poor man's" version of tracing that only works within a single binary and outputs log messages.  See [github.com/kubernetes/utils/trace](https://github.com/kubernetes/utils/tree/master/trace).
+
+Distributed tracing provides a single window into latency information from across many components and plugins. Trace data is structured, and there are numerous established backends for visualizing and querying over it.
+
+### Definitions
+
+**Span**: The smallest unit of a trace.  It has a start and end time, and is attached to a single trace.
+**Trace**: A collection of Spans which represents a single process.
+**Trace Context**: A reference to a Trace that is designed to be propagated across component boundaries.
+
+### Goals
+
+* Make it possible to visualize the progress of objects across distinct Kubernetes components
+* Streamline the tedious processes involved in debugging latency issues in Kubernetes
+* Make it possible to identify high-level latency regressions, and attribute them back to offending processes
+
+
+### Non-Goals
+
+* Replace existing logging, metrics, or the events API
+* Trace operations from all Kubernetes resource types in a generic manner (i.e. without manual instrumentation)
+
+## Proposal
+
+### Architecture
+
+#### Tracing API Requests
```

Comment:
Will it be possible for me to force a trace to be sampled e.g. by passing a request header with a traceid that I can search for later on? This is very useful for debugging.

Reply:
You can force a trace to be sampled with the `sampled` [`trace-flags` option](https://w3c.github.io/trace-context/#trace-flags).  That is propagated along with all of the other trace information, and tracing libraries should export the trace to be exported by all involved components.  I believe @Monkeyanator made use of this with OpenCensus while he was working on his POC.  In the future, there are definitely some neat integrations we could choose to do with kubectl to make this easy to use.

You can set the trace id as well, but i've found it to be easier to search for tags that are added to spans (e.g. Object Name, Object UID, etc.) to get at a specific trace.